harness/determined/keras/_data.py

```diff
@@ -169,198 +298,55 @@ def __init__(
-                is True, the iterator stops at the end of the epoch.
-        """
-        self._is_validation = is_validation
-        self._sequence._batch_offset = batch_offset
-        if self._workers > 0:
-            self._enqueuer = tf.keras.utils.OrderedEnqueuer(
-                self._sequence, use_multiprocessing=self._use_multiprocessing
-            )
-            self._enqueuer.start(workers=self._workers, max_queue_size=self._max_queue_size)
-
-    def get_iterator(self) -> Iterator:
-        """
-        Gets an Iterator over the data.
-
-        `start` must be called prior to calling this function"
-        """
-
-        def _make_finite(iterator: Iterator, num_steps: int) -> Iterator:
-            for _ in range(num_steps):
-                yield next(iterator)
-
-        def _iter_sequence_infinite(sequence: tf.keras.utils.Sequence) -> Iterator:
-            while True:
-                yield from sequence
-
-        if self._is_validation:
-            if self._workers > 0:
-                iterator = self._enqueuer.get()
-                return _make_finite(iterator, len(self._sequence))
-            return iter(self._sequence)
-
-        if self._workers > 0:
-            return self._enqueuer.get()  # type: ignore
-        return _iter_sequence_infinite(self._sequence)
-
-    def stop(self, timeout: Optional[int] = None) -> None:
-        """
-        Stops processing the data.
-
-        If workers is >0, this will stop any related threads and processes.
-        Otherwise this is a no-op.
-
-        Args:
-            timeout: Maximum time to wait.
-        """
-        if self._workers > 0:
-            self._enqueuer.stop(timeout=timeout)
-
-
-InputData = Union[tf.keras.utils.Sequence, tf.data.Dataset, SequenceAdapter, tuple]
 
+        self.sequence = sequence
+        self.use_multiprocessing = use_multiprocessing
+        self.workers = workers
+        self.max_queue_size = max_queue_size
 
-def _get_x_y_and_sample_weight(
-    input_data: Union[tf.keras.utils.Sequence, tf.data.Dataset, SequenceAdapter, tuple]
-) -> Tuple[Any, Any, Any]:
-    if isinstance(input_data, (tf.keras.utils.Sequence, tf.data.Dataset, SequenceAdapter)):
-        return input_data, None, None
 
-    elif isinstance(input_data, tuple) and len(input_data) == 2:
-        return input_data[0], input_data[1], None
+def _adapt_data_from_data_loader(
+    input_data: InputData,
+    batch_size: int,
+) -> Union[tf.keras.utils.Sequence, SequenceAdapter, tf.data.Dataset]:
+    if isinstance(input_data, tf.data.Dataset):
+        return input_data
 
-    elif isinstance(input_data, tuple) and len(input_data) == 3:
-        return input_data[0], input_data[1], input_data[2]
+    if isinstance(input_data, (tf.keras.utils.Sequence, SequenceAdapter)):
+        return input_data
 
-    else:
+    if not isinstance(input_data, tuple) or len(input_data) not in (2, 3):
```

Comment:
non-blocking: this doesn't seem to be easier to understand nor save lines of code. :)